### ISSUE TITLE
Website elements not getting translated

### ISSUE THREAD

**Issue by joelnitta** [swcarpentry-ja/git-novice-working#2]
Currently, the website (https://swcarpentry-ja.github.io/git-novice/) looks like this:

![image](https://github.com/user-attachments/assets/05128d29-5400-46bd-8eee-f89af2ff7811)

The navigation buttons along the top should be in Japanese, but appear in English.

When I render the translated website locally with `dovetail::render_trans_from_branch()` the website elements get translated properly, so this seems to be related to the [GitHub Action](https://github.com/swcarpentry-ja/git-novice/blob/main/.github/workflows/sandpaper-main.yaml).

**Comment by joelnitta** [comment]
I have created some test repos in order to troubleshoot this.

- This repo reproduces the problem, and is basically a copy of the current repo (git-novice):
  - site: https://joelnitta.github.io/git-novice-test/
  - source: https://github.com/joelnitta/git-novice-test

- This is a minimal lesson that includes translations but I am **unable** to reproduce the problem.
  - site: https://joelnitta.github.io/test-dovetail/
  - source: https://github.com/joelnitta/test-dovetail
  
The question is, what is the difference between the two that is causing the problem??

**Comment by froggleston** [comment]
Are there any differences in the SessionInfo() of each?

**Comment by joelnitta** [comment]
`test-dovetail` (renders correctly):

https://github.com/joelnitta/test-dovetail/actions/runs/11012301927/job/30578204739#step:8:36

```
R version 4.4.1 (2024-06-14)
Platform: x86_64-pc-linux-gnu
Running under: Ubuntu 22.04.5 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.10.0 
LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.10.0

locale:
 [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       
 [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   
 [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          
[10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   

time zone: UTC
tzcode source: system (glibc)

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

loaded via a namespace (and not attached):
 [1] processx_3.8.4   compiler_4.4.1   R6_2.5.1         assertthat_0.2.1
 [5] cli_3.6.3        tools_4.4.1      withr_3.0.1      callr_3.7.6     
 [9] ps_1.8.0         rlang_1.1.4      renv_1.0.9       sandpaper_0.16.7
 ```
 
`git-novice-test` (does not render correctly):

https://github.com/joelnitta/git-novice-test/actions/runs/11013982827/job/30583553763

```
R version 4.4.1 (2024-06-14)
Platform: x86_64-pc-linux-gnu
Running under: Ubuntu 22.04.5 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.10.0 
LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.10.0

locale:
 [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       
 [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   
 [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          
[10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   

time zone: UTC
tzcode source: system (glibc)

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

loaded via a namespace (and not attached):
 [1] processx_3.8.4   compiler_4.4.1   R6_2.5.1         assertthat_0.2.1
 [5] cli_3.6.3        tools_4.4.1      withr_3.0.1      callr_3.7.6     
 [9] ps_1.8.0         rlang_1.1.4      renv_1.0.9       sandpaper_0.16.7
```

They look the same

**Comment by froggleston** [comment]
~~Sorry if this is a naive question based on how dovetail works, but I thought the lesson navigation and other elements were translated via the Workbench using the `lang` in the config.yaml? I can't see a `lang: ja` in the config that the site is built from?~~

Ah I can see that dovetail sets the config lang and rewrites the config file. I can confirm the lesson element translation works locally for me too.

**Comment by froggleston** [comment]
The only thing I can think of is that the env variable to set the code isn't being pulled in properly?

`echo "LANG_CODE=${{ secrets.LANG_CODE }}" >> ~/.Renviron`

If each action step is run in a separate shell, then the LANG_CODE will be empty for the subsequent dovetail action step? Does dovetail read in the language from the Renviron file? Apologies for all the questions!

**Comment by tobyhodges** [comment]
Have you tried clearing the caches on the GitHub Actions, @joelnitta?

**Comment by joelnitta** [comment]
@froggleston :

> If each action step is run in a separate shell, then the LANG_CODE will be empty for the subsequent dovetail action step? 

That does not explain it, because loading LANG_CODE happens in the same step as rendering:

```
      - name: "Deploy Site"
        run: |
          reset <- "${{ github.event.inputs.reset }}" == "true"
          options(repos = c(carpentries = "https://carpentries.r-universe.dev/", CRAN = "https://cran.rstudio.com/"))
          renv::install("carpentries/sandpaper")
          renv::install("joelnitta/dovetail")
          sandpaper::package_cache_trigger(TRUE)
          lesson_trans_dir <- paste0(tempdir(), "/dovetail-", Sys.Date())
          lang_code <- Sys.getenv("LANG_CODE", unset = "en")
          sessionInfo()
          dovetail:::make_translated_dir(
            translated_dir = lesson_trans_dir, overwrite = TRUE, lang = lang_code,
            l10n_branch = NULL, clean = FALSE)
          sandpaper:::ci_deploy(path = lesson_trans_dir, reset = reset)
        shell: Rscript {0}
```

> Does dovetail read in the language from the Renviron file? Apologies for all the questions!

Yes. It's a bit circuitous, but the only way I was able to get it to load from a variable that the user specifies in GitHub settings (thus allowing us to use the same GitHub Action YAML for all repos).

**Comment by joelnitta** [comment]
@tobyhodges :

> Have you tried clearing the caches on the GitHub Actions, @joelnitta?

Yes, many times

**Comment by froggleston** [comment]
> That does not explain it, because loading LANG_CODE happens in the same step as rendering

The system env variable LANG_CODE is set in the previous step, and then accessed in the subsequent step. I'm wondering if the shell is the same shell instance between two steps in the same runner. If not, then that LANG_CODE will be unset in the Deploy Site step.

You may have to use the GITHUB_ENV properties to persist system env variables across steps...

**Comment by joelnitta** [comment]
> The system env variable LANG_CODE is set in the previous step

Yes, but that is why I write it to ~/.Renviron, then load it from there.

**Comment by froggleston** [comment]
> then load it from there.

I'm afraid I can't see where that's happening - could you point me to where this happens in the GHA/code?

I also don't understand from the GHA where the translation of the lesson elements is happening. I understand that the l10n_main branch is used for the translated content (as is used by the `make_translated_dir` function), but where does sandpaper know the language to use when translating the lesson elements? Again, apologies for my lack of knowledge here!

**Comment by froggleston** [comment]
So I have been messing around in [this repo](https://github.com/froggleston/git-novice-ja-test) and I got the elements to translate with [this config](https://github.com/froggleston/git-novice-ja-test/blob/l10n_main/locale/ja/config.yaml):

![{B4F9F5D2-01C9-4ECA-A5F7-79E59FBC211D}](https://github.com/user-attachments/assets/ceab0418-105d-41dd-9474-7c246564892d)

I changed the lesson type from `swc` to `incubator` ~~, and removed the URL and analytics options.~~

I also presume you have the LANG_CODE repo secret set:

![{0B7FEC8F-25CD-4B86-9574-25C0530BA6E0}](https://github.com/user-attachments/assets/4ba5d77e-c391-45bb-90c7-002759c1e3c5)

**Comment by joelnitta** [comment]
@froggleston 

Wow, I did not think of changing the lesson type, but I can confirm [that single change](https://github.com/joelnitta/git-novice-test/commit/3eafec220dc5c4b02cc82bf243d09f5e2e075bc5) does indeed [fix it](https://joelnitta.github.io/git-novice-test/)! So that means **something is going on within {sandpaper} and friends**, not the GHA.

I assume we actually need `swc` and not `incubator` though, so I don't think this is a final fix. (odd though that we don't see the "Pre-alpha" badge in the upper-left, I thought that got added to all incubator lessons. anyways.)

To answer your [question](https://github.com/swcarpentry-ja/git-novice/issues/2#issuecomment-2373965339)...

[Here](https://github.com/joelnitta/git-novice-test/blob/639a19aac2865893ad3961bf9fb005a06657aa1b/.github/workflows/sandpaper-main.yaml#L61) we write `LANG_CODE` from the Repository Secrets to `~/.Renviron` (yes, I specify LANG_CODE in the repository secrets)

```
echo "LANG_CODE=${{ secrets.LANG_CODE }}" >> ~/.Renviron
```

That means `LANG_CODE` will now be available as an environmental variable in any subsequent R session because of [how `.Renviron` works](https://support.posit.co/hc/en-us/articles/360047157094-Managing-R-with-Rprofile-Renviron-Rprofile-site-Renviron-site-rsession-conf-and-repos-conf).

[Then](https://github.com/joelnitta/git-novice-test/blob/639a19aac2865893ad3961bf9fb005a06657aa1b/.github/workflows/sandpaper-main.yaml#L71C11-L71C61) we assign the environmental variable to the `lang_code` object:

```
lang_code <- Sys.getenv("LANG_CODE", unset = "en")
```

**Comment by froggleston** [comment]
Thanks for the information about `.Renviron` - I wasn't aware of the level of automatic processing of env variables there.

The peculiar thing is that sandpaper works *locally* with no changes. I'm still thinking it's something to do with how GHAs access some part of the configuration at certain times....

**Comment by joelnitta** [comment]
The change in behavior between `carpentry: 'swc'` to `carpentry: 'incubator'` in `config.yaml` makes me suspect otherwise though, or at least that {sandpaper} is involved someway...

**Comment by tobyhodges** [comment]
A full log of weirdness encountered with this today ([link to my test repo](https://github.com/tobyhodges/sturdy-enigma)):

- `LANG_CODE` set to `'ja'`: content in Japanese; headings* in English
- `LANG_CODE` set to `'uk'`: content in Ukrainian; **headings in Ukrainian**
- `LANG_CODE` set to `'ja'`, `config.yaml` in the root folder of the `l10n_main` branch adjusted so that `carpentry` is set to `'incubator'`: content in Japanese; **headings in Japanese; styling still with Software Carpentry branding**
- `LANG_CODE` set to `'ja'`, `carpentry` reverted to `'swc'` in `config.yaml` but `keywords` adjusted to include an extra word, `'test'` : content in Japanese; **headings in Japanese**
- `LANG_CODE` set to `'ja'`, `config.yaml` unchanged except for a couple of extra newline characters after the `carpentry` field: content in Japanese; **headings in Japanese**
- Revert changes above and restore `config.yaml` of `l10n_main` branch to its original state: content in Japanese; headings in English again
- `LANG_CODE` set to `'ja'`, newlines added to `locale/ja/config.yaml`: content in Japanese; **headings in Japanese**

*'headings' here includes dropdown button text, etc.

In summary, Ukrainian headings work with the repository as-is. And _any change at all_ made to the `config.yaml` file in the root or the `locale/ja/` folder of the `l10n_main` branch will cause the Japanese headings to be used.

🤯

**Comment by tobyhodges** [comment]
@joelnitta based on the findings above, and comparing the incremental file lists in the build logs of [sites that did](https://github.com/tobyhodges/sturdy-enigma/actions/runs/11053727649/job/30708796154) and [did not](https://github.com/swcarpentry-ja/git-novice/actions/runs/11005181238/job/30557417182) include the headings in Japanese (plus the fact that the headings appear just fine in Ukrainian with the current repo state), @froggleston and I suspect the issue is may be with the behaviour of `rsync` in the `system2` calls in `dovetail:::make_translated_dir`. 

Please could you try playing around a bit with the options you are using with `rsync` in that function? Rob suggests that the `-t` option (bundled into `-a`) might be the troublemaker.

**Comment by joelnitta** [comment]
[Fixed !!](https://github.com/joelnitta/dovetail/pull/27) Thank you so much @froggleston and @tobyhodges , this was some amazing detective work.
